Fix display and discovery for interpreters included via settings or found in `/opt/python`

sharon-wang opened the pull request:
### Summary

- addresses https://github.com/posit-dev/positron/issues/6270
- addresses https://github.com/posit-dev/positron/issues/7058

#### Implementation

- set the environment `source` to `PythonEnvSource.UserSettings` for interpreters specified via `python.interpreters.include`/`python.interpreters.override` for both the JS and Native locators
- ensure user settings-specified interpreters are tagged as Custom instead of Unknown
    - if an interpreter was added via the settings, but is a Venv, it will show as a Venv instead of Custom; same goes for Uv/Conda/etc.
- ensure `/opt/python`-installed interpreters tagged as Global instead of Unknown

#### Known Issues
See screenshots in the QA Notes section below for examples
- when using the Native locator, the Python: Select Interpreter quickpick does not display as much information as when the JS locator is used. In particular, text like "custom" is not included 

### Release Notes

#### New Features

- Python interpreters installed at `/opt/python` or included via interpreter settings are no longer tagged as "Unknown" (#6270)

#### Bug Fixes

- Fixed issue with missing or duplicated Python interpreters when installed in `opt/python` or included via interpreter settings (#7058)


### QA Notes

@:interpreters

#### Setup
- ensure at least one python installation exists at `/opt/python`,
    - e.g. `/opt/python/3.10.4/bin/python`
- ensure at least one python installation is specified in `python.interpreters.include`
    - e.g. `~/scratch/3.10.4/bin/python` exists and the following is set:
    ```
    "python.interpreters.include": [
            "~/scratch"
    ],
    ```

#### Settings+UI Combinations

1. "python.locator": "js" and "console.multipleConsoleSessions": false

    ##### Interpreter Dropdown

    <img width="386" alt="jslocator+nomulticonsole+interpreterdropdown" src="https://github.com/user-attachments/assets/7b941ea0-1308-4012-8b81-64ac4003203f" />

    ##### Python: Select Interpreter Quickpick (`console.multipleConsoleSessions` can be true or false)

    <img width="607" alt="jslocator+pythonquickpick" src="https://github.com/user-attachments/assets/27a74920-e6ff-4cc7-a54a-0d70815b9c4e" />

2. "python.locator": "js" and "console.multipleConsoleSessions": true

    ##### Multi Console Start Session Quickpick

    <img width="612" alt="jslocator+multiconsole+startsessionquickpick" src="https://github.com/user-attachments/assets/1bbc9f68-e2a5-4f21-a46b-6c796d5e6a6f" />

3. "python.locator": "native" and "console.multipleConsoleSessions": false

    ##### Interpreter Dropdown

    <img width="383" alt="locatornative_muticonsolefalse_interpreterdropdown" src="https://github.com/user-attachments/assets/c7626328-ce33-43d4-8129-63db24ceef46" />

    ##### Python: Select Interpreter Quickpick (`console.multipleConsoleSessions` can be true or false)

    <img width="609" alt="locatornative_quickpick" src="https://github.com/user-attachments/assets/eb676ba6-9a15-4792-9382-e426c07207ab" />

4. "python.locator": "native" and "console.multipleConsoleSessions": true

    ##### Multi Console Start Session Quickpick

    <img width="610" alt="locatornative_muticonsoletrue" src="https://github.com/user-attachments/assets/70aad74d-f4c9-47fe-a8ea-8a2ad437e8f9" />

This change is also reflected in the dropdown in the Project Wizard, which retrieves the same information as the Interpreter Dropdown / Start Session Quickpick:

<img width="707" alt="image" src="https://github.com/user-attachments/assets/686ee754-0e53-4505-b586-fa6672a2e053" />

sharon-wang committed:
fix duplicate /opt/python discovery with JS locator

sharon-wang committed:
add user settings PythonEnvSource

sharon-wang committed:
include additional posix paths when checking for global envs

sharon-wang committed:
include custom python envs in Venv creation quickpick

sharon-wang committed:
add environment identifier for PythonEnvKind.Custom

sharon-wang committed:
add EnvironmentType.Custom to prioritized env types

sharon-wang committed:
group Custom env types together

for some reason, the /opt/python POSIX paths are ending up in the Custom group, even though they should be Global.

sharon-wang committed:
ensure /opt/python are tagged as global installed envs

sharon-wang committed:
show Custom interpreters right after the Recommended interpreter

sharon-wang committed:
group interpreter settings envs under their underlying env type

e.g. Uv, Venv, Global, etc.

i.e. don't show a separate group in the python: select interpreter quickpick for Custom env types

sharon-wang committed:
tag entries in interpreter quickpick as "custom" if added via user settings

sharon-wang committed:
⧛ things are working with JS locator + non-multi-console

- Python: Select Interpreter
- Top Bar Interpreter Dropdown

sharon-wang committed:
⧛ yay JS locator + multi-console is fine too

sharon-wang committed:
tag custom envs when using native locator

sharon-wang committed:
native locator: tag custom and /opt/python interpreters as such (instead of Unknown)

sharon-wang committed:
🧹 misc minor cleanup

sharon-wang committed:
move env kind fixup earlier during env resolution

sharon-wang committed:
update user specified env locator unit test

sharon-wang committed:
don't display venvs/etc with extra 'custom' tag

sharon-wang commented on extensions/positron-python/src/client/pythonEnvironments/base/locators/lowLevel/posixKnownPathsLocator.ts:
```diff
@@ -90,18 +94,13 @@ export class PosixKnownPathsLocator extends Locator<BasicEnvInfo> {
-        for await (const dir of additionalDirs) {
-            const { filename } = dir;
+export async function* getAdditionalPosixBinaries(searchDepth = 2): AsyncGenerator<string> {
+    for (const location of ADDITIONAL_POSIX_BIN_PATHS) {
+        const executables = findInterpretersInDir(location, searchDepth);
+        for await (const entry of executables) {
+            const { filename } = entry;
             if (await looksLikeBasicGlobalPython(filename)) {
-                yield path.dirname(filename);
+                yield filename;
```
This change fixes the duplicate python installations issue for the JS Locator. I wasn't sure where to apply a similar fix for the Native Locator. If anyone has guidance on this, happy to add in a fix as part of this PR; otherwise, I'll open an issue to address this separately.

![image](https://github.com/user-attachments/assets/6c1f101c-e9d8-488e-903e-6d9e2ebfcba3)

sharon-wang commented on extensions/positron-python/src/client/pythonEnvironments/base/locators/lowLevel/posixKnownPathsLocator.ts:
```diff
@@ -43,21 +42,26 @@ export class PosixKnownPathsLocator extends Locator<BasicEnvInfo> {
+
                 for (const bin of pythonBinaries) {
                     try {
                         yield { executablePath: bin, kind, source: [PythonEnvSource.PathEnvVar] };
```
Are we discovering Python installations from the `PATH` or `PYTHONPATH` environment variables?

isabelizimm replied:
We use PYTHONPATH to find other modules, but PATH to find environments. Looking at [PythonEnvSource](https://github.com/posit-dev/positron/blob/main/extensions/positron-python/src/client/pythonEnvironments/base/info/index.ts#L95) might be helpful here

- [VSCode docs on PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH)
- [Python lang PYTHONPATH docs](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH)

isabelizimm reviewed (dismissed):
lgtm, working as expected! thanks for all the details...wow is there a lot of permutations to view/switch/discover Pythons 😩

sharon-wang commented on extensions/positron-python/src/client/interpreter/configuration/interpreterSelector/commands/setInterpreter.ts:
```diff
@@ -727,6 +727,10 @@ function getGroup(item: IInterpreterQuickPickItem, workspacePath?: string) {
     }
     switch (item.interpreter.envType) {
+        // --- Start Positron ---
+        case EnvironmentType.Custom:
```
I implemented this grouping at one point, but when I looked at the dropdown, I felt like "Custom" Uv/Venv/etc. environments should be under the respective categories of Uv/Venv/etc. instead of "Custom".

I'm not 100% sure though! As a Python user, is it more helpful to know the type of environment at-a-glance (Venv, Uv, Global) or the fact that the interpreter is a Custom one added via settings?

I figure one can always check their settings if they want to know if an interpreter is Custom-added, but there are less places to check at-a-glance what environment type their interpreter is.

isabelizimm replied:
My thought process: if I have specifically added these interpreters, it's probably because I _really_ want to use them. It would be slightly easier to discover if all of them were grouped together. However, it's probably a more confusing failure mode if someone has a venv that they're expecting under the venv grouping and can't find. 

Let's leave it your way for now! We can always update it later if people feel strongly about it.

sharon-wang committed:
simplify reduction of user-specified interpreters into a single interpreter per directory

sharon-wang committed:
fix isCustomEnvironment check

Check if any custom environment directories are parents of the interpreter path, instead of just checking if the interpreter path is specified in the include or override settings.

This is needed because the resolved interpreter path may not be the same as the one specified in the settings.

e.g. ~/scratch/3.10.4/bin/python3 may be resolved to ~/scratch/3.10.4/bin/python because we prioritize the basic python executable.

sharon-wang committed:
filter out duplicate native locator entries for additional env dirs

sharon-wang committed:
move duplicate env handling to a separate function

sharon-wang committed:
js locator: filter out duplicate posix paths more correctly

sharon-wang committed:
native locator: filter out duplicate additional paths more correctly

sharon-wang committed:
js locator: filter out duplicate user specified paths more correctly

sharon-wang committed:
💅

sharon-wang committed:
clarify additional dir existing env handling

sharon-wang committed:
💅
